xarray/core/accessor_str.py

```diff
@@ -76,17 +80,46 @@ class StringAccessor:
     """
 
     __slots__ = ("_obj",)
+    _pattern_type = type(re.compile(""))
 
     def __init__(self, obj):
         self._obj = obj
 
-    def _apply(self, f, dtype=None):
+    def _apply(self, f, obj=None, dtype=None, otype=None, signature=None, **kwargs):
         # TODO handling of na values ?
+        if obj is None:
+            obj = self._obj
         if dtype is None:
-            dtype = self._obj.dtype
+            dtype = obj.dtype
+        if otype is None:
+            otype = dtype
```

Comment:
What is the difference between `dtype` and `otype` should they not be the same?

Reply:
Ok I see below why but what is `result.dtype` when `dtype` and `otype` differ? Consider the following example:

```python
import numpy as np
import xarray as xr

da = xr.DataArray(["a", "b", "cd"])
print(da.dtype)

g = np.vectorize(lambda x: x.upper(), otypes=[np.object_])
xr.apply_ufunc(g, da, output_dtypes=[da.dtype]).dtype # -> is object
```

so it seems there is no point in making a difference between `dtype` and `otype` - am I wrong?

Reply:
I didn't realize that `assert_equal` wasn't comparing dtypes.  So you are right, this was returning a different dtype.  I fixed this and simplified the code.